lib/engines/vllm0_8/src/lib.rs

```diff
@@ -0,0 +1,557 @@
+// SPDX-FileCopyrightText: Copyright (c) 2024-2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::ffi::CString;
+use std::path::PathBuf;
+use std::str::FromStr;
+use std::time::Duration;
+use std::{path::Path, sync::Arc};
+
+use async_stream::stream;
+use dynamo_llm::engines::MultiNodeConfig;
+use dynamo_llm::protocols::common::FinishReason;
+use dynamo_runtime::engine::{AsyncEngine, AsyncEngineContextProvider, ResponseStream};
+use dynamo_runtime::pipeline::error as pipeline_error;
+use dynamo_runtime::pipeline::{async_trait, Error, ManyOut, SingleIn};
+use dynamo_runtime::protocols::annotated::Annotated;
+use dynamo_runtime::{CancellationToken, Result};
+use pyo3_async_runtimes::TaskLocals;
+use serde::Deserialize;
+use tokio::sync::oneshot::Sender;
+
+use dynamo_llm::backend::ExecutionContext;
+use dynamo_llm::protocols::common::llm_backend::{BackendInput, LLMEngineOutput};
+use pyo3::prelude::*;
+use pyo3::types::{IntoPyDict, PyDict, PyString, PyTuple};
+use pyo3::PyObject;
+use pyo3::Python;
+use pythonize::pythonize;
+use tokio_stream::StreamExt;
+
+// The minor revision version of vllm that this engine supports
+const VLLM_VERSION: &str = "0.8";
+
+const PY_MAIN_MOD: &str = include_str!("vllm_inc.py");
+
+const VLLM_SHUTDOWN_TIMEOUT: Duration = Duration::from_secs(3);
+
+pub async fn make_engine(
+    cancel_token: CancellationToken,
+    // Full path to the model, either a GGUF file or an HF repo dir
+    model_path: &Path,
+    // Multi node settings
+    node_conf: MultiNodeConfig,
+    // How many GPUs to use
+    tensor_parallel_size: u32,
+    // Path to extra engine args file
+    extra_engine_args: Option<PathBuf>,
+) -> pipeline_error::Result<ExecutionContext> {
+    let engine = VllmV1Engine::new(
+        cancel_token,
+        model_path,
+        node_conf,
+        tensor_parallel_size,
+        extra_engine_args,
+    )
+    .await?;
+    let engine: ExecutionContext = Arc::new(engine);
+    Ok(engine)
+}
+
+struct VllmV1Engine {
+    cancel_token: CancellationToken,
+    // How we send requests to Python / vllm
+    request_queue: Arc<PyObject>,
+    // asyncio event loop to run all our python futures. vllm is async.
+    event_loop: Arc<PyObject>,
+    // The python module from vllm_inc.py
+    py_main_mod: Arc<PyObject>,
+    // vllm.SamplingParams
+    sampling_params: PyObject,
+}
+
+impl VllmV1Engine {
```

Comment:
Have you tried this with both `VLLM_USE_V1=0` and `VLLM_USE_V1=1`?

Reply:
vllm says this on startup:
> > [arg_utils.py:1742] Engine in background thread is experimental on VLLM_USE_V1=1. Falling back to V0 Engine

vllm (`asyncio`, really) wants to own the main thread, but so does Rust's `tokio`, so currently it's v0 only. The good news is that if the experimental check gets removed it will just work on v1.

I renamed that class to `VllmEngine` to reflect that.